agent/worker.go

```diff
@@ -484,3 +488,40 @@ func (w *worker) Subscribe(ctx context.Context, subscription *api.SubscriptionMe
 		}
 	}
 }
+
+// Leave removes all assignments and waits for task exit
+func (w *worker) Leave(ctx context.Context) error {
+	var err error
+
+	var tms []*taskManager
+	w.mu.RLock()
+	for _, tm := range w.taskManagers {
+		tms = append(tms, tm)
+	}
+	w.mu.RUnlock()
+
+	if err = w.Assign(ctx, nil); err != nil {
+		return err
+	}
+
+	// wait for taskManagers close
+	var wg sync.WaitGroup
+	for _, tm := range tms {
+		wg.Add(1)
+		go func(tm *taskManager) {
+			defer wg.Done()
+			select {
+			case <-tm.closed:
```

Comment:
I think you can just call `tm.Close()` here. Otherwise, you risk coupling the workers and task managers too tightly.